### ISSUE TITLE
[API] No way to obtain image pull progress

### ISSUE THREAD

**Issue by jelly** [containers/podman#12341]
/kind feature

**Description**

In cockpit-podman we are currently showing pull progress as "pulling container from XYZ", it would be nicer to shows the progress as a percentage.

The current `/libpod/images/pull'` API uses a POST to pull the image and returns once the download is complete, the stream status is only returned after completion.

I can understand that the current behaviour of the API can't be changed as that would break compatibility, maybe an event can be emitted during downloading?

As containers are usually downloaded pretty quickly due to it's usually small size it's not a huge UX issue, but would be a nice improvement to have it as an option.

**Comment by Luap99** [comment]
Are you sure that it is not streaming? If I pull a big image with curl I definitely see the stream output right away.

**Comment by jelly** [comment]
```
[jelle@t14s][/tmp]%curl -XPOST --unix-socket $XDG_RUNTIME_DIR/podman/podman.sock "http://d/v3.0.0/libpod/images/pull?reference=docker.io/grafana/grafana"
{"stream":"Trying to pull docker.io/grafana/grafana:latest...\n"}
{"stream":"Getting image source signatures\n"}
{"stream":"Copying blob sha256:f73d9d8d4dd9e2a9f052acd4de025cc3fd927e366877b54194bd07520d8115cd\n"}
{"stream":"Copying blob sha256:ac27e49d9e31bb3f9d11aad63d4c914048bc691f945c92e376f68a5ebcac5b8e\n"}
{"stream":"Copying blob sha256:a0d0a0d46f8b52473982a3c466318f479767577551a53ffc9074c9fa7035982e\n"}
{"stream":"Copying blob sha256:0ae21741e86c18ac8abc3dfecd8ed8d0c6b1edb4248c64ebac211f4a50a0069d\n"}
{"stream":"Copying blob sha256:a40f2a35c016154c70b04bb5b5a259e9d9e6be132ba5d20d11325f5770e98f25\n"}
{"stream":"Copying blob sha256:b55eac8c5e31448cc9f1415577d9311ebdd002e6a6d3b7558d8fb0faa1f0db6b\n"}
{"stream":"Copying blob sha256:a40f2a35c016154c70b04bb5b5a259e9d9e6be132ba5d20d11325f5770e98f25\n"}
{"stream":"Copying blob sha256:0ae21741e86c18ac8abc3dfecd8ed8d0c6b1edb4248c64ebac211f4a50a0069d\n"}
{"stream":"Copying blob sha256:f73d9d8d4dd9e2a9f052acd4de025cc3fd927e366877b54194bd07520d8115cd\n"}
{"stream":"Copying blob sha256:ac27e49d9e31bb3f9d11aad63d4c914048bc691f945c92e376f68a5ebcac5b8e\n"}
{"stream":"Copying blob sha256:a0d0a0d46f8b52473982a3c466318f479767577551a53ffc9074c9fa7035982e\n"}
{"stream":"Copying blob sha256:b55eac8c5e31448cc9f1415577d9311ebdd002e6a6d3b7558d8fb0faa1f0db6b\n"}
{"stream":"Copying blob sha256:4f4fb700ef54461cfa02571ae0db9a0dc1e0cdb5577484a6d75e68dc38e8acc1\n"}
{"stream":"Copying blob sha256:ccd7ded5cb2a5446698bad905c7def57e06270d166d9815d14ccb461df1bd5b0\n"}
{"stream":"Copying blob sha256:a88d683c5ad24002b28771724b3321880506ec98c640939d5c0d074cf9ababbd\n"}
{"stream":"Copying blob sha256:4f4fb700ef54461cfa02571ae0db9a0dc1e0cdb5577484a6d75e68dc38e8acc1\n"}
{"stream":"Copying blob sha256:ccd7ded5cb2a5446698bad905c7def57e06270d166d9815d14ccb461df1bd5b0\n"}
{"stream":"Copying blob sha256:a88d683c5ad24002b28771724b3321880506ec98c640939d5c0d074cf9ababbd\n"}
{"stream":"Copying config sha256:ddfae340d0681fe1a10582b06a2e8ae402196df9d429f0c1cefbe8dedca73cf0\n"}
{"stream":"Writing manifest to image destination\n"}
{"stream":"Storing signatures\n"}
{"images":["ddfae340d0681fe1a10582b06a2e8ae402196df9d429f0c1cefbe8dedca73cf0"],"id":"ddfae340d0681fe1a10582b06a2e8ae402196df9d429f0c1cefbe8dedca73cf0"}
[jelle@t14s][/tmp]%
```

Indeed shows the progress, but a frontend can't present this as a download progress in terms of percentage downloaded.

**Comment by Luap99** [comment]
@vrothberg @mtrmac @jwhonce PTAL

**Comment by mtrmac** [comment]
c/image does provide a (somewhat) more detailed progress API for Go callers; compare https://github.com/cri-o/cri-o/blob/407d866343dfa3fca4997089894d05b15c5a64b3/server/image_pull.go#L174-L211 . I don’t immediately know whether it is sufficient to produce the Docker progress stream.

**Comment by rhatdan** [comment]
@vrothberg @mtrmac @jwhonce any movement on this issue?

**Comment by gbraad** [comment]
This also affects our usecase for the Podman console view in the CRC application. People reported they aren't informed about the download progress.

**Comment by gbraad** [comment]
Still relevant

**Comment by msisj** [comment]
I'm also interested in this feature ([stated in another podman repo - podman-py](https://github.com/containers/podman-py/pull/173#issuecomment-1090404837)).

@rhatdan @jwhonce @vrothberg Are there any plans on adding this feature ? I can see that in https://github.com/containers/podman/issues/11554 (kind of similar request, but podman-remote push rather than REST pull) similar feature was requested 8 months ago.

**Comment by vrothberg** [comment]
Thank you for reaching our, @msisj!

There are definitely plans on adding the feature.  It is mostly a question of priority and time.

**Comment by doehyunbaek** [comment]
Hi. I wish to tackle this issue because at my job I deal with huge size images(2GB+).

Can anyone give a guidance on what feature is needed on what part of podman ecosystem? I'm not sure whether it  needs some changes to containers/image repo or changes in podman repo would suffice.

**Comment by mtrmac** [comment]
> Can anyone give a guidance on what feature is needed on what part of podman ecosystem?  I'm not sure whether it needs some changes to containers/image repo or changes in podman repo would suffice.

See if https://github.com/containers/image/blob/e594853d6471bc9ab7e8231588d3199f38c75322/copy/copy.go#L133 provides enough data for Podman’s progress stream. If not, c/image changes might be necessary.

`c/image/copy.Image` is called from `c/common/libimage`, so that might need to be updated at least enough to pass the data through unmodified.

And finally, there’s Podman itself, with a somewhat deep call stack between the command line → (API client → API server) → local c/common caller. Podman experts can make better recommendations than me in this respect.

**Comment by doehyunbaek** [comment]
Thanks for your pointer, I would tinker with that file and how it's called from podman to see if I can progress. Really appreciate it.

If I get stuck I'll update here.

**Comment by doehyunbaek** [comment]
Hi, during work I found out that podman indeed do give progress as percentage and was confused if this was something requested here.

https://user-images.githubusercontent.com/70570804/169806744-8cda92b9-7c4c-4a26-be13-9a44dbe2bd80.mov

Maybe this feature is already included in podman? my podman version is

```console
$ podman version
Client:       Podman Engine
Version:      4.1.0
API Version:  4.1.0
Go Version:   go1.18
Built:        Sat May  7 01:15:54 2022
OS/Arch:      linux/amd64
```

**Comment by garrett** [comment]
@doehyunbaek: This is about the Podman API, not the command line interface. Pull status is needed for software that uses Podman via the API, such as [Cockpit-Podman](https://github.com/cockpit-project/cockpit-podman), which is a graphical, web-based way to use Podman on servers through [Cockpit](https://cockpit-project.org/).

**Comment by doehyunbaek** [comment]
Thanks for the reply, will look into Podman API and if pull progress exposed in cli can be also expoed at API.

**Comment by mtrmac** [comment]
@doehyunbaek The `c/image/copy.Image` code produces both those progress bars (just as a byte stream to a file descriptor), and the `ProgressProperties` data (as a Go data structure). The progress bar byte stream can’t be reconstructed into the requested API-consumable form.

**Comment by doehyunbaek** [comment]
Thanks for the reply.

I think I should explore existing ways of communicating progress through http and then experiment with changing podman api serving code / manipulating underlying ProgressProperties data to suit my need.

**Comment by jmguzik** [comment]
In the scope of https://github.com/containers/podman/pull/14886 I implemented a solution similar to the docker one, but there is another option that has to be explored for which I, unfortunately, don't have time right now. The attached PR should provide more information. cc @vrothberg

**Comment by gbraad** [comment]
The CLI does, but not the API

**Comment by sochoaAtMulesoft** [comment]
It would seem that CLI on MacOS Ventura also does not have progress bars to indicate download progress.

**Comment by mtrmac** [comment]
@sochoaAtMulesoft This issue is specifically about the API, please file a separate one if you are reporting something different.

**Comment by vrothberg** [comment]
@jelly @garrett would the [data provided by the c/image library](https://github.com/containers/image/blob/e594853d6471bc9ab7e8231588d3199f38c75322/copy/copy.go#L133) do the job for you?

The pull stream can be extended to include such  a `types.ProgressProperties` which includes details on the pull/push progress of each blob (layers, configs, manifests, etc.) which can in turn be used to render progress bars.

**Comment by jelly** [comment]
@vrothberg that sounds good, do I understand it correctly that it would look something alike:

```
{"stream":"Copying blob sha256:f73d9d8d4dd9e2a9f052acd4de025cc3fd927e366877b54194bd07520d8115cd\n", progress: 10}
{"stream":"Copying blob sha256:ac27e49d9e31bb3f9d11aad63d4c914048bc691f945c92e376f68a5ebcac5b8e\n", progress: 15}
```

If there is already a PR I or a patch somewhere I'm happy to build podman locally and give it a try :+1:

**Comment by mtrmac** [comment]
@vrothberg Just to be sure this is mentioned, compare https://github.com/containers/podman/pull/14968 implementing progress in the compat API.

---

> do I understand it correctly that it would look something alike:
>
> {"stream":"Copying blob sha256:f73d9d8d4dd9e2a9f052acd4de025cc3fd927e366877b54194bd07520d8115cd\n", progress: 10}

I imagine an ideal long-term API should contain data fields and not (English, unlocalized) text. It should _especially_ not motivate consumers to _parse text_ to extract data fields, because then we could not change the text in any way.

(OTOH a free-flow text _is_ a good way to make an API future-proof, exactly because more data and more kinds of data can be added to text without breaking API — _only_ as long as everyone agrees that the text is not to be parsed by machines.)

What does “progress” mean here? Is it a global percentage? Bytes offset in that blob? Percentage in that blob? (All three could be _possibly_ useful… with per-blob data only useful if a blob identifier is also provided.)

**Comment by vrothberg** [comment]
> What does “progress” mean here? Is it a global percentage? Bytes offset in that blob? Percentage in that blob? (All three could be possibly useful… with per-blob data only useful if a blob identifier is also provided.)

It would be a [types.ProgressProperties](https://github.com/containers/image/blob/e594853d6471bc9ab7e8231588d3199f38c75322/copy/copy.go#L133).  This way, old clients checking for `.Stream != ""` can continue working and will discard the new messages with `.Progress`.

**Comment by benoitf** [comment]
I think one missing part is the number of layers (as we discover them when a new one is pulled) so would be nice to have the information as soon as possible as well

And missing part of Podman vs Docker is the progressDetail field

in Docker
```
 {"status":"Downloading","progressDetail":{"current":5626202,"total":29152458},"progress":"[=========\u003e                                         ]  5.626MB/29.15MB","id":"3ae0c06b4d3a"}
```

in Podman
```
progressDetail":{}
```

**Comment by vrothberg** [comment]
> I think one missing part is the number of layers (as we discover them when a new one is pulled) so would be nice to have the information as soon as possible as well

I don't think Docker has this kind of information, does it?

> And missing part of Podman vs Docker is the progressDetail field...

The data returned by the Podman endpoint would be different than the one of Docker (or the compat endpoint). Podman uses [types.ProgressProperties](https://github.com/containers/image/blob/e594853d6471bc9ab7e8231588d3199f38c75322/copy/copy.go#L133), so I think that's what the endpoints should expose.

**Comment by benoitf** [comment]
>>I think one missing part is the number of layers (as we discover them when a new one is pulled) so would be nice to have the information as soon as possible as well

>I don't think Docker has this kind of information, does it?

No, but you asked about data we may need ;-)

>> And missing part of Podman vs Docker is the progressDetail field...

> The data returned by the Podman endpoint would be different than the one of Docker (or the compat endpoint). Podman >uses [types.ProgressProperties]
>(https://github.com/containers/image/blob/e594853d6471bc9ab7e8231588d3199f38c75322/copy/copy.go#L133), so I
> think that's what the endpoints should expose.


well Podman should still provide these data in the compat endpoint for all tools using the socket as a docker socket. (as it is missing now) so they'll get progress (and if I use docker cli using a remote podman socket, etc)

**Comment by benoitf** [comment]
> please open another issue for this one. Here, I want to focus on the libpod/Podman endpoint.

Ok I will because your original question was 'the data you need for rendering in Podman Desktop?' and Podman Desktop is using the compat endpoint when pulling images

**Comment by vrothberg** [comment]
@baude @mtrmac @rhatdan @Luap99 this is issue (and https://github.com/containers/podman/issues/12341) has an elevated priority for Desktop.  Please take a look and share what you think.

My current proposal:
* Add a new parameter to libpod's pull endpoint.  When set, it will stream [types.ProgressProperties](https://github.com/containers/image/blob/e594853d6471bc9ab7e8231588d3199f38c75322/copy/copy.go#L133) instead. 
* Initially, there will be no plumbing on the remote-client side.  Ideally, the progress bars in c/image/copy would be updated at some point to also use [types.ProgressProperties](https://github.com/containers/image/blob/e594853d6471bc9ab7e8231588d3199f38c75322/copy/copy.go#L133) such that Linux-native and remote clients can share the very same code for rendering progress bars.
* Once that is done, we can add a similar streaming mechanism to the kube/play endpoint.

I discussed yesterday with @benoitf that Desktop can - in theory - already show pull progress.  Desktop uses the compat endpoint for rendering progress when doing an image pull.  If Desktop pre-processes the K8s YAML and pulls the specified images _before_ firing the kube/play endpoint, it would achieve what's desired https://github.com/containers/podman/issues/15902.

Tackling this issue here and exposing progress on the libpod pull endpoint is something that needs to happen nonetheless.

**Comment by vrothberg** [comment]
> Ok I will because your original question was 'the data you need for rendering in Podman Desktop?' and Podman Desktop is using the compat endpoint when pulling images

The data structures to render progress returned by `kube/play` and the libpod's pull endpoint will likely be different than the one of the compat endpoint.

**Comment by mtrmac** [comment]
> this issue (and #12341)

@vrothberg _this_ is #12341; did you intend to link to something else?

---

> * Add a new parameter to libpod's pull endpoint.  When set, it will stream [types.ProgressProperties](https://github.com/containers/image/blob/e594853d6471bc9ab7e8231588d3199f38c75322/copy/copy.go#L133) instead.

This is possible but not something I’d _enthusiastically_ recommend. `ProgressProperties` is, at this point, not a really comprehensively designed feature (it basically only cares about individual layers, so as @benoitf says, there’s nothing about how many layers in an image there are, or how many per-platform instances in a multi-platform image). IIRC it’s only the minimum that was necessary for one specific caller, growing only a tiny bit over time for other users.

And those missing pieces, to an extent, already show that the `ProgressProperties` type is not _obviously_ an ideal way to represent the alternative “subclasses of a progress event superclass” (e.g. per-layer data vs. non-layer data.)


OTOH, using `ProgressProperties` (with an external, Podman-defined, JSON representation?) or something very similar would not be a disaster — with the proposed new parameter to opt into this data, we could, in the future, add a “progressV2” parameter that opts into data of some _other_ form.

So I can imagine that exporting `ProgressProperties` more or less as is, and not trying to design a close-to-perfect final data type, at the cost of possibly having to maintain two implementations long-term, might be a good trade-off. I don’t have a particularly strong opinion.


> * Initially, there will be no plumbing on the remote-client side.  Ideally, the progress bars in c/image/copy would be updated at some point to also use [types.ProgressProperties](https://github.com/containers/image/blob/e594853d6471bc9ab7e8231588d3199f38c75322/copy/copy.go#L133) such that Linux-native and remote clients can share the very same code for rendering progress bars.

That would be valuable and exciting for the Podman remote CLI (and it would clearly require extending `ProgressProperties` quite a bit). Podman Desktop probably doesn’t immediately care.

**Comment by mtrmac** [comment]
I think _ideally_ we (by which I mean $someoneUnknown) should think (and work with Podman Desktop) to define what events, and what data about those events, are relevant to consumers.

The tickets already contain a fair amount of specific requests. At some meta-level:
- Do we define a very specific data structure giving callers all data they need, and then force them to deal with the complexity?
- Do we, instead, focus on just “providing a progress bar that advances”, and define some kind of model for that? (Possibly a hierarchy of actions - “podman play kube” → create pod → create container → pull layers → layer progress; with each action having a percentage scale or just action count/total.)
- Both?
- Something else?
- Do we also provide pre-formatted text for TUI in the stream? Or a (Go?) library to turn that text into the TUI from the detailed data, so that remote Podman has exactly the same output as local?

The Docker-provided API should definitely be considered.

---

There’s definitely a lot of opportunities to architecture-astronaut this and turn it into an impossible project. Hence I’m not being enthusiastic about exposing `ProgressProperties` as is, but I’m also not strongly opposed.

**Comment by vrothberg** [comment]
> I think ideally we (by which I mean $someoneUnknown) should think (and work with Podman Desktop) to define what events, and what data about those events, are relevant to consumers.

I think we've done that already and gravitated toward progress (bars) for image pulls only.

> The Docker-provided API should definitely be considered.

I share that feeling as well.  Podman has the code to convert `ProgressProperties` to the Docker JSON and Desktop has the code to render that data.

**Comment by mtrmac** [comment]
I’m sorry, I meant “… be considered as a source of inspiration”. I didn’t look into the details of that API enough to say whether it should be adopted as it exists right now.

**Comment by vrothberg** [comment]
Very related issue is https://github.com/containers/podman/issues/15714 which highlights that podman-remote pull doesn't render progress bars.

**Comment by vrothberg** [comment]
@baude @mtrmac I created https://issues.redhat.com/browse/RUN-1963  as a SPIKE to follow up on the libpod-specific work

**Comment by vrothberg** [comment]
I think we can close this issue now.

https://github.com/containers/podman/issues/15714 tracks the parts of plumbing in podan-remote.
